### PR TITLE
feat: Add storing output as json

### LINKED ISSUE #1437: Save flank output as json file
**Author the user story for this feature**

As a Flank user, I want to have the output saved to the file so I can easily parse this file on CI or use it in Flank integration tests.

**Describe the solution**

1. Add command ```--output-file-format``` to determine the output file type.
1. Save output file to matrix directory
1. Upload to gcs
1. Use it in integration tests it could help with the easiest tests update

What team think about this proposal?
@bootstraponline @piotradamczyk5 @jan-gogo @pawelpasterz @Sloox @axelzuziak-gogo

### PR DESCRIPTION
**Pull request by piotradamczyk5**
Fixes #1437 

1. Enable output report using option ( cli: `--output-report=json`, yaml: `output-report: json` under `flank` root)
1. Run Flank and see that the output reported is stored ( file `outputReport.json`)
1. Compare file content with console output

## Checklist

- [x] [SDD](https://github.com/Flank/flank/issues/1437#issuecomment-769299621)
- [x] Documented
- [x] Unit tested
- [x] Integration tests updated

### CODE REVIEW

File: test_runner/src/main/kotlin/ftl/reports/output/OutputReportConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    enabled = enableOutputReport,
+    local = OutputReportLocalConfiguration(localStorageDirectory),
+    remote = OutputReportRemoteConfiguration(
+        uploadReport = disableResultsUpload.not(),
+        resultsBucket = resultsBucket,
+        resultsDir = resultsDir
+    )
+)
+
+private val IArgs.localStorageDirectory
```
Comment (adamfilipow92) [inline test_runner/src/main/kotlin/ftl/reports/output/OutputReportConfiguration.kt:33]: It could be reusable, wdyt about moving this property to IArgs.kt?

Comment (piotradamczyk5) [reply test_runner/src/main/kotlin/ftl/reports/output/OutputReportConfiguration.kt:33]: moved 👍

Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/reports/output/OutputReportConfiguration.kt:33]: Thanks! I will use it in my pr 😄

**Comment by piotradamczyk5** [comment]
@mergifyio rebase

**Review by jan-gogo — APPROVED** [review]
## Test Report
### Run
From repository root
```bash
cd test_runner
flank run android
```
### Result
`outputReport.json`:
```json
{
  "args": {
    "commonArgs": {
      "data": "---\ngcloud:\n  app: \"../test_projects/android/apks/app-debug.apk\"\n  test: \"../test_projects/android/apks/app-debug-androidTest.apk\"\nflank: {}\n",
      "devices": [
        {
          "model": "NexusLowRes",
          "version": "28",
          "locale": "en",
          "orientation": "portrait",
          "isVirtual": true
        }
      ],
      "resultsBucket": "test-lab-v9cn46bb990nx-kz69ymd4nm9aq",
      "resultsDir": "2021-02-04_16-48-31.143000_OhrI",
      "recordVideo": false,
      "testTimeout": "15m",
      "async": false,
      "flakyTestAttempts": 0,
      "otherFiles": {},
      "directoriesToPull": [],
      "scenarioNumbers": [],
      "failFast": false,
      "project": "flank-open-source",
      "maxTestShards": 1,
      "shardTime": -1,
      "repeatTests": 1,
      "smartFlankGcsPath": "",
      "smartFlankDisableUpload": false,
      "testTargetsAlwaysRun": [],
      "filesToDownload": [],
      "disableSharding": false,
      "localResultDir": "results",
      "runTimeout": "-1",
      "fullJUnitResult": false,
      "ignoreFailedTests": false,
      "keepFilePath": false,
      "outputStyle": "Verbose",
      "disableResultsUpload": false,
      "defaultTestTime": 120.0,
      "defaultClassTestTime": 240.0,
      "useAverageTestTimeForNewTests": false,
      "disableUsageStatistics": false,
      "outputReportType": "JSON"
    },
    "appApk": "/Users/jan-gogo/projects/flank-project/flank/test_projects/android/apks/app-debug.apk",
    "testApk": "/Users/jan-gogo/projects/flank-project/flank/test_projects/android/apks/app-debug-androidTest.apk",
    "additionalApks": [],
    "autoGoogleLogin": false,
    "useOrchestrator": true,
    "roboDirectives": [],
    "environmentVariables": {},
    "grantPermissions": "all",
    "scenarioLabels": [],
    "obbFiles": [],
    "obbNames": [],
    "performanceMetrics": false,
    "testTargets": [],
    "additionalAppTestApks": [],
    "useLegacyJUnitResult": false,
    "obfuscateDumpShards": false,
    "testTargetsForShard": []
  },
  "cost": {
    "physical": 0.00,
    "virtual": 0.02,
    "total": 0.02
  },
  "test_results": {
    "matrix-1g0ipqet4iz4m": [
      {
        "device": "NexusLowRes-28-en-portrait",
        "outcome": "success",
        "details": "1 test cases passed",
        "testSuiteOverview": {
          "total": 1,
          "errors": 0,
          "failures": 0,
          "flakes": 0,
          "skipped": 0,
          "elapsedTime": 1.075,
          "overheadTime": 0.0
        }
      }
    ]
  },
  "weblinks": [
    "https://console.firebase.google.com/project/flank-open-source/testlab/histories/bh.bf178d418be9d33e/matrices/7143386027735431817/executions/bs.6aa4fb040cccf3bb"
  ]
}
```
### Summary
* The implementation looks good 👍 
* `outputResult.json` looks as expected

Good Job!

**Comment by jan-gogo** [comment]
We need one review more and test different config

**Comment by adamfilipow92** [comment]
Please update pr description

**Review by adamfilipow92 — APPROVED** [review]
- [X] Code looks good
- [X] Output works - tested for ios ```simple-ios-flank``` with ```output-report``` flag
- [X] Integration tests works, tested locally (macos)

Here is my ```outputReport.json```
```
{
  "args": {
    "useLegacyJUnitResult": true,
    "xcTestRunData$delegate": {
      "_value": {
        "rootDir": "/Users/adamfilipowicz/Repos/flank/test_runner/src/test/kotlin/ftl/fixtures/tmp/ios/EarlGreyExample/",
        "nsDict": {
          "EarlGreyExampleSwiftTests": {
            "BlueprintName": {
              "content": "EarlGreyExampleSwiftTests"
            },
            "BundleIdentifiersForCrashReportEmphasis": {
              "array": [
                {
                  "content": "io.gogoapps.earlgrey.samples.EarlGreySwiftTests"
                },
                {
                  "content": "io.gogoapps.earlgrey.samples.EarlGrey"
                }
              ]
            },
            "CommandLineArguments": {
              "array": []
            },
            "DefaultTestExecutionTimeAllowance": {
              "type": 0,
              "longValue": 600,
              "doubleValue": 600.0,
              "boolValue": false
            },
            "DependentProductPaths": {
              "array": [
                {
                  "content": "__TESTROOT__/Debug-iphoneos/EarlGreyExampleSwift.app"
                },
                {
                  "content": "__TESTROOT__/Debug-iphoneos/EarlGreyExampleSwift.app/PlugIns/EarlGreyExampleSwiftTests.xctest"
                }
              ]
            },
            "EnvironmentVariables": {
              "DYLD_INSERT_LIBRARIES": {
                "content": "@executable_path/EarlGrey.framework/EarlGrey"
              },
              "OS_ACTIVITY_DT_MODE": {
                "content": "YES"
              },
              "SQLITE_ENABLE_THREAD_ASSERTIONS": {
                "content": "1"
              }
            },
            "IsAppHostedTestBundle": {
              "type": 2,
              "longValue": 1,
              "doubleValue": 1.0,
              "boolValue": true
            },
            "ProductModuleName": {
              "content": "EarlGreyExampleTestsSwift"
            },
            "RunOrder": {
              "type": 0,
              "longValue": 0,
              "doubleValue": 0.0,
              "boolValue": false
            },
            "SystemAttachmentLifetime": {
              "content": "deleteOnSuccess"
            },
            "TestBundlePath": {
              "content": "__TESTHOST__/PlugIns/EarlGreyExampleSwiftTests.xctest"
            },
            "TestHostBundleIdentifier": {
              "content": "io.gogoapps.earlgrey.samples.EarlGrey"
            },
            "TestHostPath": {
              "content": "__TESTROOT__/Debug-iphoneos/EarlGreyExampleSwift.app"
            },
            "TestLanguage": {
              "content": ""
            },
            "TestRegion": {
              "content": ""
            },
            "TestTimeoutsEnabled": {
              "type": 2,
              "longValue": 0,
              "doubleValue": 0.0,
              "boolValue": false
            },
            "TestingEnvironmentVariables": {
              "DYLD_INSERT_LIBRARIES": {
                "content": "__TESTHOST__/Frameworks/libXCTestBundleInject.dylib"
              },
              "XCInjectBundleInto": {
                "content": "unused"
              }
            },
            "ToolchainsSettingValue": {
              "array": []
            },
            "UITargetAppCommandLineArguments": {
              "array": []
            },
            "UserAttachmentLifetime": {
              "content": "deleteOnSuccess"
            }
          },
          "__xctestrun_metadata__": {
            "FormatVersion": {
              "type": 0,
              "longValue": 1,
              "doubleValue": 1.0,
              "boolValue": false
            }
          }
        },
        "version": "V1",
        "shardTargets": {
          "": [
            {
              "EarlGreyExampleSwiftTests": [
                "EarlGreyExampleSwiftTests/testWithGreyAssertions",
                "EarlGreyExampleSwiftTests/testLayout",
                "EarlGreyExampleSwiftTests/testThatThrows",
                "EarlGreyExampleSwiftTests/testWithInRoot",
                "EarlGreyExampleSwiftTests/testCustomAction",
                "EarlGreyExampleSwiftTests/testWithCondition",
                "EarlGreyExampleSwiftTests/testBasicSelection",
                "EarlGreyExampleSwiftTests/testWithCustomMatcher",
                "EarlGreyExampleSwiftTests/testCollectionMatchers",
                "EarlGreyExampleSwiftTests/testCatchErrorOnFailure",
                "EarlGreyExampleSwiftTests/testWithCustomAssertion",
                "EarlGreyExampleSwiftTests/testTableCellOutOfScreen",
                "EarlGreyExampleSwiftTests/testBasicSelectionAndAction",
                "EarlGreyExampleSwiftTests/testBasicSelectionAndAssert",
                "EarlGreyExampleSwiftTests/testWithCustomFailureHandler",
                "EarlGreyExampleSwiftTests/testBasicSelectionActionAssert",
                "EarlGreyExampleSwiftTests/testSelectionOnMultipleElements"
              ]
            }
          ]
        },
        "shardChunks": {
          "": [
            {
              "testMethodNames": [
                "EarlGreyExampleSwiftTests/testWithGreyAssertions",
                "EarlGreyExampleSwiftTests/testLayout",
                "EarlGreyExampleSwiftTests/testThatThrows",
                "EarlGreyExampleSwiftTests/testWithInRoot",
                "EarlGreyExampleSwiftTests/testCustomAction",
                "EarlGreyExampleSwiftTests/testWithCondition",
                "EarlGreyExampleSwiftTests/testBasicSelection",
                "EarlGreyExampleSwiftTests/testWithCustomMatcher",
                "EarlGreyExampleSwiftTests/testCollectionMatchers",
                "EarlGreyExampleSwiftTests/testCatchErrorOnFailure",
                "EarlGreyExampleSwiftTests/testWithCustomAssertion",
                "EarlGreyExampleSwiftTests/testTableCellOutOfScreen",
                "EarlGreyExampleSwiftTests/testBasicSelectionAndAction",
                "EarlGreyExampleSwiftTests/testBasicSelectionAndAssert",
                "EarlGreyExampleSwiftTests/testWithCustomFailureHandler",
                "EarlGreyExampleSwiftTests/testBasicSelectionActionAssert",
                "EarlGreyExampleSwiftTests/testSelectionOnMultipleElements"
              ],
              "testMethods": [
                {
                  "name": "EarlGreyExampleSwiftTests/testWithGreyAssertions",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testLayout",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testThatThrows",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testWithInRoot",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testCustomAction",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testWithCondition",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testBasicSelection",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testWithCustomMatcher",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testCollectionMatchers",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testCatchErrorOnFailure",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testWithCustomAssertion",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testTableCellOutOfScreen",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testBasicSelectionAndAction",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testBasicSelectionAndAssert",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testWithCustomFailureHandler",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testBasicSelectionActionAssert",
                  "time": 120.0,
                  "isParameterized": false
                },
                {
                  "name": "EarlGreyExampleSwiftTests/testSelectionOnMultipleElements",
                  "time": 120.0,
                  "isParameterized": false
                }
              ]
            }
          ]
        }
      }
    },
    "commonArgs": {
      "data": "---\ngcloud:\n  test: \"./test_runner/src/test/kotlin/ftl/fixtures/tmp/ios/EarlGreyExample/EarlGreyExample.zip\"\n  xctestrun-file: \"./test_runner/src/test/kotlin/ftl/fixtures/tmp/ios/EarlGreyExample/EarlGreyExampleSwiftTests.xctestrun\"\n  results-dir: \"test_dir\"\nflank:\n  output-report: \"json\"\n",
      "devices": [
        {
          "model": "iphone8",
          "version": "13.6",
          "locale": "en",
          "orientation": "portrait",
          "isVirtual": false
        }
      ],
      "resultsBucket": "test-lab-v9cn46bb990nx-kz69ymd4nm9aq",
      "resultsDir": "test_dir",
      "recordVideo": false,
      "testTimeout": "15m",
      "async": false,
      "flakyTestAttempts": 0,
      "otherFiles": {},
      "type": "XCTEST",
      "directoriesToPull": [],
      "scenarioNumbers": [],
      "failFast": false,
      "project": "flank-open-source",
      "maxTestShards": 1,
      "shardTime": -1,
      "repeatTests": 1,
      "smartFlankGcsPath": "",
      "smartFlankDisableUpload": false,
      "testTargetsAlwaysRun": [],
      "filesToDownload": [],
      "disableSharding": false,
      "localResultDir": "results",
      "runTimeout": "-1",
      "fullJUnitResult": false,
      "ignoreFailedTests": false,
      "keepFilePath": false,
      "outputStyle": "Single",
      "disableResultsUpload": false,
      "defaultTestTime": 120.0,
      "defaultClassTestTime": 240.0,
      "useAverageTestTimeForNewTests": false,
      "disableUsageStatistics": false,
      "outputReportType": "JSON"
    },
    "xctestrunZip": "/Users/adamfilipowicz/Repos/flank/test_runner/src/test/kotlin/ftl/fixtures/tmp/ios/EarlGreyExample/EarlGreyExample.zip",
    "xctestrunFile": "/Users/adamfilipowicz/Repos/flank/test_runner/src/test/kotlin/ftl/fixtures/tmp/ios/EarlGreyExample/EarlGreyExampleSwiftTests.xctestrun",
    "testTargets": [],
    "obfuscateDumpShards": false,
    "additionalIpas": [],
    "app": "",
    "testSpecialEntitlements": false,
    "onlyTestConfiguration": "",
    "skipTestConfiguration": ""
  },
  "cost": {
    "physical": 0.08,
    "virtual": 0.00,
    "total": 0.08
  },
  "test_results": {
    "matrix-1iimg89acphkj": [
      {
        "device": "iphone8-13.6-en-portrait",
        "outcome": "success",
        "details": "17 test cases passed",
        "testSuiteOverview": {
          "total": 17,
          "errors": 0,
          "failures": 0,
          "flakes": 0,
          "skipped": 0,
          "elapsedTime": 22.419,
          "overheadTime": 0.0
        }
      }
    ]
  },
  "weblinks": [
    "https://console.firebase.google.com/project/flank-open-source/testlab/histories/bh.a3b607c9bb6d0088/matrices/7683345571758085610/executions/bs.9a78518a6501c50e"
  ]
}
```

**Pull request merged**